db/db_impl.cc

```diff
@@ -5117,16 +5117,29 @@ Status DBImpl::DelayWrite(uint64_t num_bytes,
   bool delayed = false;
   {
     StopWatch sw(env_, stats_, WRITE_STALL, &time_delayed);
-    auto delay = write_controller_.GetDelay(env_, num_bytes);
+    uint64_t stall_start = env_->NowNanos();
+    uint64_t delay = write_controller_.GetDelay(env_, num_bytes);
     if (delay > 0) {
       if (write_options.no_slowdown) {
         return Status::Incomplete();
       }
+      TEST_SYNC_POINT_CALLBACK("DBImpl::DelayWrite:Sleep", &delay);
+
       mutex_.Unlock();
-      delayed = true;
-      TEST_SYNC_POINT("DBImpl::DelayWrite:Sleep");
-      // hopefully we don't have to sleep more than 2 billion microseconds
-      env_->SleepForMicroseconds(static_cast<int>(delay));
+      // We will delay the write until we have slept for delay ms or
+      // we don't need a delay anymore
+      const uint64_t kDelayInterval = 10000;
+      uint64_t stall_end = stall_start + (delay * std::milli::den);
+      while (write_controller_.NeedsDelay()) {
+        if (env_->NowNanos() >= stall_end) {
```

Comment:
I didn't mean we need to call a monotonic clock here. I just want to avoid TimedWait() if possible. But what you are doing here is still correct.